### ISSUE TITLE
Do not remove unused packages from straight lockfile

### ISSUE THREAD

**Issue by mohkale** [radian-software/straight.el#1096]
Hi there,

In my Emacs config I often have several optional packages which I don't even install if I haven't enabled them for my current environment. It seems like straight removes the package version from the lockfile if I run straight-freeze-versions while that package is not available. Then when I re-enable the package it auto upgrades to the latest version. Would it be possible to retain packages whose versions have not changed in the lockfile instead of deleting them as if they've been permanently removed?

**Comment by raxod502** [comment]
This is the purpose of `straight-register-package`. The intent is that if you have a conditional package, you always register, but may not invoke `straight-use-package`. It will then continue to be tracked in your configuration and lockfile, but not compiled or added to `load-path`.

**Comment by mohkale** [comment]
I don't think this is very practical with the way my config is setup. I basically have to generate a complete list of all dependencies ahead of time and then register them. That or update each individual package so alongside an optionally evaluated use-package block I have an explicit straight-register block. Do you have any suggestions for other projects might be going about this :thinking:. I'm aware doom just seems to lock in the use-package block directly but is there a less grindy approach you could recommend. Personally I don't see why this wouldn't behave just like the straight repos directory. You only remove things when the user chooses to prune them out, otherwise keep everything in a stable state.

**Comment by raxod502** [comment]
Well, the current implementation assumes that all packages are available in order to compute dependencies and the lockfile. We could add support for not pruning the lockfile as aggressively, but it wouldn't be a full solve: the lockfile generation would still prune undeclared packages.

Doom essentially implements its own package management layer on top of `straight.el`. This isn't really what I recommend, but I can't fault the author; `straight.el` is lacking the features that are needed to implement what Doom wants.

> That or update each individual package so alongside an optionally evaluated use-package block I have an explicit straight-register block.

You could configure your `use-package` so that it evaluates the `straight-register-package` form unconditionally, but only `straight-use-package` if your conditional form passes. That should be reasonably straightforward, I think.

I understand why this feels like a hack to you: it's because the current implementation of `straight.el` depends on some assumptions about what data is available. I have no objection to changing that, but it requires a fair bit of work which I have not (as of yet) undertaken to make the conceptual model more powerful so it can support more intelligent lockfile management. Nonetheless, I think the conditional `straight-register-package` approach is the best thing that exists today.
